src/ttlv/format.rs

```diff
@@ -0,0 +1,434 @@
+        let [header_slot, value_slot, ..] = buffer else {
+            return Err(TruncationError);
+        };
+
+        header_slot.write(header.to_be_bytes());
+        value_slot.write(((value as u32 as u64) << 32).to_be_bytes());
+        self.position += 2;
+        Ok(FormatDone(()))
+    }
+
+    /// Format a long integer.
+    #[inline]
+    pub fn format_long_int(&mut self, tag: Tag, value: i64) -> FormatResult {
+        let header = (tag.with_type(Type::LongInteger).value() as u64) << 32 | 8;
+
+        let buffer = &mut self.buffer[self.position..];
+        let [header_slot, value_slot, ..] = buffer else {
+            return Err(TruncationError);
+        };
+
+        header_slot.write(header.to_be_bytes());
+        value_slot.write((value as u64).to_be_bytes());
+        self.position += 2;
+        Ok(FormatDone(()))
+    }
+
+    /// Format a big integer.
+    ///
+    /// `value` should be the bytewise digits of a signed (two's complement)
+    /// big-endian integer.  It will be padded (with leading sign-extension
+    /// bytes) into a multiple of 8 bytes before being stored.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the length of `value` (when rounded up to a multiple of 8
+    /// bytes) does not fit in a `u32`.
+    #[inline]
+    pub fn format_big_int(&mut self, tag: Tag, value: &[u8]) -> FormatResult {
+        let num_blocks = value.len().div_ceil(8);
+        let buffer = &mut self.buffer[self.position..];
+        if buffer.len() < 1 + num_blocks {
+            return Err(TruncationError);
+        }
+
+        let header = tag.with_type(Type::BigInteger).value();
+        // NOTE: 'buffer.len() * 8 <= 2^32' so 'num_blocks * 8 < 2^32'.
+        let length = (value.len() as u32).next_multiple_of(8);
+        let header = (header as u64) << 32 | length as u64;
+        buffer[0].write(header.to_be_bytes());
+
+        if let &[first, ..] = value {
+            // Fill the first 8 bytes by sign extension.
+            //
+            // This also ensures all bytes are initialized.
+            //
+            // On the one hand, we could extract the top bit from 'first' and
+            // sign-extend just that.  On the other hand, we know that at least
+            // one byte of 'buffer[1]' (which always includes the last byte)
+            // will get overwritten by the later copy.
+            //
+            // This code is likely constant-time, which may be beneficial in
+            // case secret data is being serialized.  But we do not provide any
+            // public guarantees about constant-time operation.
+            let ext = first as i8 as i64;
+            buffer[1].write(ext.to_be_bytes());
+
+            // Fill the remaining bytes.
+            let ptr = buffer[1..].as_mut_ptr_range().end.cast::<u8>();
+            unsafe {
+                ptr.sub(value.len())
+                    .copy_from_nonoverlapping(value.as_ptr(), value.len())
+            };
```

Comment:
If (and that's a big IFF) I understand this correctly:
- `as_mut_ptr_range().end()` returns a pointer to the byte AFTER the last byte of the given range.
- The given range is `buffer[1..]` which ends at the end of `buffer`.
- As `buffer` is set above to `self.buffer[self.position..]` the end of `buffer` is the end +1 of the entire buffer given to the formatter.
- `ptr.sub(value.len())` will move the pointer to the end of the formatter buffer + 1 back by `value.len()` bytes then `copy_from_nonoverlapping()` will copy bytes into the formatter buffer at that point.

Why does this appear to be writing from the end of the _*entire*_ formatter buffer minus the value length, instead of from the current write position in the formatter buffer onwards?

What did I miss?